internal/eval/eval.go

```diff
@@ -4191,6 +4191,107 @@ func evalHRANDFIELD(args []string, store *dstore.Store) []byte {
 	return selectRandomFields(hashMap, count, withValues)
 }
 
+func evalJSONARRINDEX(args []string, store *dstore.Store) []byte {
```

Comment:
Few things to improve here:
1. we should standardise the handling of the negative indices logic here and extract it into a separate function that can be used by other methods too (please find another eval method which implements the duplicated logic and replace it with a call to the function i am writing below this.
2. Please return error messages which are exactly the same as what redis would return. The messages being returned here are inaccurate.
3. Use reflect.DeepEqual for json object comparisons. Using `==` would firstly not always give the correct response, and in the worst case can also lead to runtime panic.
4. Please add test cases (ideally as unit tests) to validate all the error paths (which cover all the three points above).

I am presenting a cleaner way to structure this code, you will however have to make a few changes to it to ensure that the error messages are consistent with redis:
```go
func evalJSONARRINDEX(args []string, store *dstore.Store) []byte {
    if len(args) < 3 || len(args) > 5 {
        return diceerrors.NewErrArity("JSON.ARRINDEX")
    }

    key := args[0]
    path := args[1]
    valueArg := args[2]

    var start, stop int
    var err error

    // Default start and stop indices
    start = 0
    stop = -1 // -1 indicates end of array

    // Parse start and stop if provided
    if len(args) >= 4 {
        start, err = strconv.Atoi(args[3])
        if err != nil {
            return diceerrors.NewErrWithMessage("invalid start index")
        }
    }

    if len(args) == 5 {
        stop, err = strconv.Atoi(args[4])
        if err != nil {
            return diceerrors.NewErrWithMessage("invalid stop index")
        }
    }

    // Retrieve the object from the database
    obj := store.Get(key)
    if obj == nil {
        return clientio.RespNIL
    }

    // Check if the object is of type JSON
    errWithMessage := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeJSON, object.ObjEncodingJSON)
    if errWithMessage != nil {
        return errWithMessage
    }

    jsonData := obj.Value

    // Check if the path specified is valid
    expr, err := jp.ParseString(path)
    if err != nil {
        return diceerrors.NewErrWithMessage("invalid JSONPath")
    }

    // Parse the value argument as JSON
    var value interface{}
    err = sonic.Unmarshal([]byte(valueArg), &value)
    if err != nil {
        // If not valid JSON, treat it as string
        value = valueArg
    }

    results := expr.Get(jsonData)
    arrIndexList := make([]interface{}, 0, len(results))

    for _, result := range results {
        if utils.GetJSONFieldType(result) == utils.ArrayType {
            arr := result.([]interface{})
            length := len(arr)

            adjustedStart, adjustedStop := adjustIndices(start, stop, length)
            if adjustedStart == -1 {
                arrIndexList = append(arrIndexList, -1)
                continue
            }

            elementFound := false
            for i := adjustedStart; i <= adjustedStop; i++ {
                if reflect.DeepEqual(arr[i], value) {
                    arrIndexList = append(arrIndexList, i)
                    elementFound = true
                    break
                }
            }
            if !elementFound {
                arrIndexList = append(arrIndexList, -1)
            }
        } else {
            arrIndexList = append(arrIndexList, nil)
        }
    }

    return clientio.Encode(arrIndexList, false)
}

// adjustIndices adjusts the start and stop indices for array traversal.
// It handles negative indices and ensures they are within the array bounds.
func adjustIndices(start, stop, length int) (int, int) {
    if length == 0 {
        return -1, -1
    }

    if start < 0 {
        start += length
    }
    if stop < 0 || stop == -1 {
        stop += length
    }

    if start < 0 {
        start = 0
    }
    if stop < 0 {
        stop = 0
    }

    if start >= length {
        return -1, -1
    }
    if stop >= length {
        stop = length - 1
    }

    if start > stop {
        return -1, -1
    }

    return start, stop
}

```